### DISCUSSION TITLE
touchlistener for callbacks

### DISCUSSION THREAD

**Discussion by Laul** [patrykandpatrick/vico#574] [Questions]
Hey ! 
New question for you ;) 

Here is roughly the context, keeping in mind that I use the Compose M3 implementation : 
* I created a screen with 2 graphs: 
  * one as 2 stacked Columns + 1 Line with only points. This represent a set of data for each day of the week
  * One composed of 1 line only. It shows the same data than the first one, but displayed as daily data.
* The second graph is intended to show a zoomed view of the first one.
* My goals are: 
  1. to be able to click on a column in the first chart and automatically scroll to the proper position on the second chart. 
  2. Ideally, it would also support the other way, ie when I scroll on the second chart, the corresponding column / day is selected on the first chart .

I found a touchevent listener in the view implementation of your lib, but nothing for Compose, except in the markervisibilitylistener. However, this function does not let me make the link between 2 different charts. 

Would you have any advice / idea to do what I'm trying to implement ?

Thanks a lot !

**Comment by patrickmichalik** [comment]
Hello! I’m afraid this isn’t possible yet, but we’re looking into expanding the API so that charts can be linked like this. Vico 2.0.0 Alpha 9 will introduce several new features that will be needed here. I’ll keep you updated.

**Comment by Laul** [comment]
Thanks for the answer! 
Maybe a simpler one : is there any way to have the control on the scroll, even if not from another chart? This is also asked by somebody else in issue #585  . If I am right we can only autoscroll at the end and start of the chart. If not directly in the lib, would you have recommendations to reach that goal ?
Thanks !

**Reply by patrickmichalik** [reply]
Yup, this is possible, @Laul. I believe I should start by clarifying the terminology. Automatic scrolling automatically changes the scroll position on data changes. This is different from programmatic scrolling, which lets you directly update the scroll position at any time.

Significant improvements have been made to both of these in [Vico 2.0.0 Alpha 9](https://github.com/patrykandpatrick/vico/releases/tag/v2.0.0-alpha.9), and automatic scrolling is no longer limited to two options, but since you appear to be referring to programmatic scrolling, let me focus on that. Pixel-based programmatic scrolling has been available since version 1.6.2. In Vico 2.0.0 Alpha 8 and earlier, such scrolling is performed via the `scrollBy` and `animateScrollBy` functions of `ChartScrollState`. [Vico 2.0.0 Alpha 9](https://github.com/patrykandpatrick/vico/releases/tag/v2.0.0-alpha.9) adds the ability to scroll to a specific _x_ value and by a specific _x_ delta, among other things. For example, this scrolls to _x_ = 10 with an animation and centers it:

```kt
scrollState.animateScroll(Scroll.Absolute.x(x = 10f, bias = .5f))
```

Pixel-based scrolling remains possible, of course—see `Scroll.Absolute.pixels` and `Scroll.Relative.pixels`.

The addition described above is the main one that I was referring to in my previous comment. The next improvement that will be needed here is better selection support—see #465. You can detect when an entry is tapped via `MarkerVisibilityChangeListener` (possibly using an empty `Marker`), and you could use this with `Scroll.Absolute.x` to link two charts, but the problem is that there’s no way of distinguishing between a selection tap and a tap that comes at the start of a scroll. Thus, the behavior of the linkage would be suboptimal. We plan on introducing a dedicated selection system.

**Comment by Gowsky** [comment]
[Vico v2.3.0-alpha.2](https://github.com/patrykandpatrick/vico/releases/tag/v2.3.0-alpha.2) introduces `CartesianMarkerController`.

**Comment by Gowsky** [comment]
That should cover everything needed for the functionality in question.

For the first point:

> to be able to click on a column in the first chart and automatically scroll to the proper position on the second chart.

With `CartesianMarkerController.toggleOnPress()`, scroll-starting presses won't be misinterpreted as selection taps. After applying it, you can use `CartesianMarkerVisibilityListener` and `Scroll.x` to set up the linkage, as @patrickmichalik described, and it will work as expected. (The dedicated selection system is still planned. With `CartesianMarkerController`, it's no longer strictly necessary for your use case, but it will make it simpler.)

For the second point:

> Ideally, it would also support the other way, ie when I scroll on the second chart, the corresponding column / day is selected on the first chart 

Regarding checking the second chart's _x_-range, see [this comment](https://github.com/patrykandpatrick/vico/discussions/984#discussioncomment-11741604). As described there, it's already possible by retrieving this data as a side effect, and we plan on making `ZoomState` directly expose this in the future. As for highlighting columns in the first chart, you can use `ColumnCartesianLayer.ColumnProvider`.

Thus, I'll be closing this discussion as resolved. Please stay tuned under the relevant feature requests for updates on the improvements mentioned.
